### PR TITLE
Issue 3412: Table Segment Metrics

### LINKED ISSUE #3412: TableSegment Metrics
**Problem description**
Need metrics for:
- put
- remove
- get
- iterators

**Problem location**
Segment Store/PravegaRequestProcessor.

**Suggestions for an improvement**

Add metrics

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**  
- Added metrics for all Table Segment Operations in PravegaRequestProcessor.

**Purpose of the change**  
Fixes #3412.

**What the code does**  
- Tracking elapsed time for table segment creation and deletion
- Tracking number of entries added and elapsed time for updates/conditional updates.
- Tracking number of keys removed and elapsed time for removals/conditional removals.
- Tracking number of items returned and elapsed time for retrievals and iterators.

**How to verify it**  
Unit tests must pass.

### CODE REVIEW

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleMonitor.java [hunk 1]
```diff
@@ -27,25 +27,29 @@
...
         this.processor = new AutoScaleProcessor(configuration, this.executor);
-        this.recorder = new SegmentStatsRecorderImpl(this.processor, store, this.executor);
+        this.statsRecorder = new SegmentStatsRecorderImpl(this.processor, store, this.executor);
+        this.tableSegmentStatsRecorder = new TableSegmentStatsRecorderImpl();
     }
 
     @Override
     public void close() {
-        this.recorder.close();
+        this.statsRecorder.close();
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleMonitor.java]: Good one! Fixed.

File: shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java [hunk 2]
```diff
@@ -56,12 +56,30 @@
...
     public static final String SEGMENT_CREATE_LATENCY = "segmentstore.segment.create_latency_ms"; // Histogram
+    public static final String SEGMENT_DELETE_LATENCY = "segmentstore.segment.delete_latency_ms"; // Histogram
     public static final String SEGMENT_READ_LATENCY = "segmentstore.segment.read_latency_ms";     // Histogram
     public static final String SEGMENT_WRITE_LATENCY = "segmentstore.segment.write_latency_ms";   // Histogram
     public static final String SEGMENT_READ_BYTES = "segmentstore.segment.read_bytes";            // Counter and Per-segment Counter
     public static final String SEGMENT_WRITE_BYTES = "segmentstore.segment.write_bytes";          // Counter and Per-segment Counter
     public static final String SEGMENT_WRITE_EVENTS = "segmentstore.segment.write_events";        // Counter and Per-segment Counter
 
+    // Table Segment stats
+    public static final String TABLE_SEGMENT_UPDATE_LATENCY = "segmentstore.tablesegment.update_latency_ms";
```
Comment (fpj) [inline shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java]: To be consistent with the other metric names in this class, could you please add a comment to the end of each new line stating the metric type?

File: shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java [hunk 3]
```diff
@@ -56,12 +56,30 @@
...
+    public static final String TABLE_SEGMENT_UPDATE_CONDITIONAL_LATENCY = "segmentstore.tablesegment.update_conditional_latency_ms";
+    public static final String TABLE_SEGMENT_REMOVE_LATENCY = "segmentstore.tablesegment.remove_latency_ms";
+    public static final String TABLE_SEGMENT_REMOVE_CONDITIONAL_LATENCY = "segmentstore.tablesegment.remove_conditional_latency_ms";
+    public static final String TABLE_SEGMENT_GET_KEYS_LATENCY = "segmentstore.tablesegment.get_latency_ms";
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java]: I fixed all of these to be consistent.

File: segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessorTest.java [hunk 4]
```diff
@@ -604,187 +617,219 @@ public void testUnimplementedMethods() throws Exception {
...
         InOrder order = inOrder(connection);
-        PravegaRequestProcessor processor = new PravegaRequestProcessor(store, tableStore, connection);
+        val recorderMock = mock(TableSegmentStatsRecorder.class);
+        PravegaRequestProcessor processor = new PravegaRequestProcessor(store, tableStore, connection, SegmentStatsRecorder.noOp(),
+                recorderMock, new PassingTokenVerifier(), false);
 
         //Generate keys
         ArrayList<HashedArray> keys = generateKeys(3, rnd);
 
         // Execute and Verify createSegment calling stack is executed as design.
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessorTest.java]: I did not write nor did I touch this line.

File: segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessorTest.java [hunk 5]
```diff
@@ -556,24 +561,32 @@ public void testUnsupportedOperation() throws Exception {
...
         StreamSegmentStore store = serviceBuilder.createStreamSegmentService();
         TableStore tableStore = serviceBuilder.createTableStoreService();
         ServerConnection connection = mock(ServerConnection.class);
         InOrder order = inOrder(connection);
-        PravegaRequestProcessor processor = new PravegaRequestProcessor(store, tableStore, connection);
+        val recorderMock = mock(TableSegmentStatsRecorder.class);
+        PravegaRequestProcessor processor = new PravegaRequestProcessor(store, tableStore, connection, SegmentStatsRecorder.noOp(),
+                recorderMock, new PassingTokenVerifier(), false);
 
         // Execute and Verify createTableSegment calling stack is executed as design.
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessorTest.java]: This one either.

File: shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java (hunk 2 above)
Comment (andreipaduroiu) [reply shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java]: Ok.

**Pull request merged**